File: binding/src/svg/libqsvggenerator.cpp
#include <QIODevice>
#include <QPaintDevice>
#include <QPaintEngine>
#include <QRect>
#include <QRectF>
#include <QSize>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <QSvgGenerator>
#include <qsvggenerator.h>
#include "libqsvggenerator.h"
#include "libqsvggenerator.hxx"

QSvgGenerator* QSvgGenerator_new() {
	 return new VirtualQSvgGenerator();
}

QSvgGenerator* QSvgGenerator_new2(int version) {
	 return new VirtualQSvgGenerator(static_cast<QSvgGenerator::SvgVersion>(version));
}

libqt_string QSvgGenerator_Title(const QSvgGenerator* self) {
	QString _ret = self->title();
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

void QSvgGenerator_SetTitle(QSvgGenerator* self, const libqt_string title) {
	self->setTitle(QString::fromUtf8(title.data, title.len));
}

libqt_string QSvgGenerator_Description(const QSvgGenerator* self) {
	QString _ret = self->description();
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

void QSvgGenerator_SetDescription(QSvgGenerator* self, const libqt_string description) {
	self->setDescription(QString::fromUtf8(description.data, description.len));
}

QSize* QSvgGenerator_Size(const QSvgGenerator* self) {
	return new QSize(self->size());
}

void QSvgGenerator_SetSize(QSvgGenerator* self, const QSize* size) {
	self->setSize(*size);
}

QRect* QSvgGenerator_ViewBox(const QSvgGenerator* self) {
	return new QRect(self->viewBox());
}

QRectF* QSvgGenerator_ViewBoxF(const QSvgGenerator* self) {
	return new QRectF(self->viewBoxF());
}

void QSvgGenerator_SetViewBox(QSvgGenerator* self, const QRect* viewBox) {
	self->setViewBox(*viewBox);
}

void QSvgGenerator_SetViewBox2(QSvgGenerator* self, const QRectF* viewBox) {
	self->setViewBox(*viewBox);
}

libqt_string QSvgGenerator_FileName(const QSvgGenerator* self) {
	QString _ret = self->fileName();
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

void QSvgGenerator_SetFileName(QSvgGenerator* self, const libqt_string fileName) {
	self->setFileName(QString::fromUtf8(fileName.data, fileName.len));
}

QIODevice* QSvgGenerator_OutputDevice(const QSvgGenerator* self) {
	return self->outputDevice();
}

void QSvgGenerator_SetOutputDevice(QSvgGenerator* self, QIODevice* outputDevice) {
	self->setOutputDevice(outputDevice);
}

void QSvgGenerator_SetResolution(QSvgGenerator* self, int dpi) {
	self->setResolution(dpi);
}

int QSvgGenerator_Resolution(const QSvgGenerator* self) {
	return self->resolution();
}

int QSvgGenerator_SvgVersion(const QSvgGenerator* self) {
	return self->svgVersion();
}

// Derived class handler implementation
QPaintEngine* QSvgGenerator_PaintEngine(const QSvgGenerator* self) {
	auto* vqsvggenerator = dynamic_cast<const VirtualQSvgGenerator*>(self);
	if (vqsvggenerator && vqsvggenerator->isVirtualQSvgGenerator) {
	return vqsvggenerator->paintEngine();
	} else {
	return self->QSvgGenerator::paintEngine();
}
}

// Base class handler implementation
QPaintEngine* QSvgGenerator_QBasePaintEngine(const QSvgGenerator* self) {
	auto* vqsvggenerator = dynamic_cast<const VirtualQSvgGenerator*>(self);
	if (vqsvggenerator && vqsvggenerator->isVirtualQSvgGenerator) {
vqsvggenerator->setQSvgGenerator_PaintEngine_IsBase(true);
	return vqsvggenerator->paintEngine();
}
}

// Auxiliary method to allow providing re-implementation
void QSvgGenerator_OnPaintEngine(const QSvgGenerator* self, intptr_t slot) {
	auto* vqsvggenerator = dynamic_cast<const VirtualQSvgGenerator*>(self);
	if (vqsvggenerator && vqsvggenerator->isVirtualQSvgGenerator) {
vqsvggenerator->setQSvgGenerator_PaintEngine_Callback(reinterpret_cast<VirtualQSvgGenerator::QSvgGenerator_PaintEngine_Callback>(slot));
}
}

// Derived class handler implementation
int QSvgGenerator_Metric(const QSvgGenerator* self, int metric) {
	auto* vqsvggenerator = dynamic_cast<const VirtualQSvgGenerator*>(self);
	if (vqsvggenerator && vqsvggenerator->isVirtualQSvgGenerator) {
	return vqsvggenerator->metric(static_cast<QPaintDevice::PaintDeviceMetric>(metric));
	} else {
	return self->QSvgGenerator::metric(static_cast<QPaintDevice::PaintDeviceMetric>(metric));
}
}

// Base class handler implementation
int QSvgGenerator_QBaseMetric(const QSvgGenerator* self, int metric) {
	auto* vqsvggenerator = dynamic_cast<const VirtualQSvgGenerator*>(self);
	if (vqsvggenerator && vqsvggenerator->isVirtualQSvgGenerator) {
vqsvggenerator->setQSvgGenerator_Metric_IsBase(true);
	return vqsvggenerator->metric(static_cast<QPaintDevice::PaintDeviceMetric>(metric));
}
}

// Auxiliary method to allow providing re-implementation
void QSvgGenerator_OnMetric(const QSvgGenerator* self, intptr_t slot) {
	auto* vqsvggenerator = dynamic_cast<const VirtualQSvgGenerator*>(self);
	if (vqsvggenerator && vqsvggenerator->isVirtualQSvgGenerator) {
vqsvggenerator->setQSvgGenerator_Metric_Callback(reinterpret_cast<VirtualQSvgGenerator::QSvgGenerator_Metric_Callback>(slot));
}
}

void QSvgGenerator_Delete(QSvgGenerator* self) {
    delete self;
}

File: binding/src/svg/libqsvggenerator.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");%%_IMPORTLIBS_%%
%%_STRUCTDEFS_%%

/// https://doc.qt.io/qt-6/qsvggenerator.html
pub const qsvggenerator = struct {

    /// New constructs a new QSvgGenerator object.
    pub fn New() QtC.QSvgGenerator {
        return qtc.QSvgGenerator_new();
    }


    /// New2 constructs a new QSvgGenerator object.
    pub fn New2(version: i32) QtC.QSvgGenerator {
        return qtc.QSvgGenerator_new2(@intCast(version));
    }


    pub fn Title(self: ?*anyopaque, , allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QSvgGenerator_Title(@ptrCast(self));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qsvggenerator.Title: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn SetTitle(self: ?*anyopaque, title: []const u8) void {
        const title_str = qtc.libqt_string{
    .len = title.len,
    .data = title.ptr,
};
qtc.QSvgGenerator_SetTitle(@ptrCast(self), title_str);
    }

    pub fn Description(self: ?*anyopaque, , allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QSvgGenerator_Description(@ptrCast(self));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qsvggenerator.Description: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn SetDescription(self: ?*anyopaque, description: []const u8) void {
        const description_str = qtc.libqt_string{
    .len = description.len,
    .data = description.ptr,
};
qtc.QSvgGenerator_SetDescription(@ptrCast(self), description_str);
    }

    pub fn Size(self: ?*anyopaque, ) QtC.QSize {
        return qtc.QSvgGenerator_Size(@ptrCast(self));
    }

    pub fn SetSize(self: ?*anyopaque, size: ?*anyopaque) void {
        qtc.QSvgGenerator_SetSize(@ptrCast(self), @ptrCast(size));
    }

    pub fn ViewBox(self: ?*anyopaque, ) QtC.QRect {
        return qtc.QSvgGenerator_ViewBox(@ptrCast(self));
    }

    pub fn ViewBoxF(self: ?*anyopaque, ) QtC.QRectF {
        return qtc.QSvgGenerator_ViewBoxF(@ptrCast(self));
    }

    pub fn SetViewBox(self: ?*anyopaque, viewBox: ?*anyopaque) void {
        qtc.QSvgGenerator_SetViewBox(@ptrCast(self), @ptrCast(viewBox));
    }

    pub fn SetViewBox2(self: ?*anyopaque, viewBox: ?*anyopaque) void {
        qtc.QSvgGenerator_SetViewBox2(@ptrCast(self), @ptrCast(viewBox));
    }

    pub fn FileName(self: ?*anyopaque, , allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QSvgGenerator_FileName(@ptrCast(self));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qsvggenerator.FileName: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn SetFileName(self: ?*anyopaque, fileName: []const u8) void {
        const fileName_str = qtc.libqt_string{
    .len = fileName.len,
    .data = fileName.ptr,
};
qtc.QSvgGenerator_SetFileName(@ptrCast(self), fileName_str);
    }

    pub fn OutputDevice(self: ?*anyopaque, ) QtC.QIODevice {
        return qtc.QSvgGenerator_OutputDevice(@ptrCast(self));
    }

    pub fn SetOutputDevice(self: ?*anyopaque, outputDevice: ?*anyopaque) void {
        qtc.QSvgGenerator_SetOutputDevice(@ptrCast(self), @ptrCast(outputDevice));
    }

    pub fn SetResolution(self: ?*anyopaque, dpi: i32) void {
        qtc.QSvgGenerator_SetResolution(@ptrCast(self), dpi);
    }

    pub fn Resolution(self: ?*anyopaque, ) i32 {
        return qtc.QSvgGenerator_Resolution(@ptrCast(self));
    }

    pub fn SvgVersion(self: ?*anyopaque, ) i32 {
        return qtc.QSvgGenerator_SvgVersion(@ptrCast(self));
    }

    pub fn PaintEngine(self: ?*anyopaque, ) QtC.QPaintEngine {
        return qtc.QSvgGenerator_PaintEngine(@ptrCast(self));
    }

    /// Allows for overriding the related default method
    pub fn OnPaintEngine(self: ?*anyopaque, callback: *const fn (?*anyopaque) callconv(.c) QtC.QPaintEngine) void {
        qtc.QSvgGenerator_OnPaintEngine(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBasePaintEngine(self: ?*anyopaque, ) QtC.QPaintEngine {
        return qtc.QSvgGenerator_QBasePaintEngine(@ptrCast(self));
    }

    pub fn Metric(self: ?*anyopaque, metric: i32) i32 {
        return qtc.QSvgGenerator_Metric(@ptrCast(self), @intCast(metric));
    }

    /// Allows for overriding the related default method
    pub fn OnMetric(self: ?*anyopaque, callback: *const fn (?*anyopaque, i32) callconv(.c) i32) void {
        qtc.QSvgGenerator_OnMetric(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseMetric(self: ?*anyopaque, metric: i32) i32 {
        return qtc.QSvgGenerator_QBaseMetric(@ptrCast(self), @intCast(metric));
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QSvgGenerator_Delete(@ptrCast(self));
    }
};

/// https://doc.qt.io/qt-6/svggenerator.html#types
pub const enums = struct {
    pub const SvgVersion = enum {
        pub const SvgTiny12: i32 = 0;
        pub const Svg11: i32 = 1;
    };

};
